Logs: Improve plugins developer documentation for logs APIs (#72441)
grafana/grafana: docs/sources/developers/plugins/create-a-grafana-plugin/extend-a-plugin/add-log-queries.md

In the case where the underlying database does not return an `id` field, you can implement one within the data source. For example, in the Loki data source, a combination of nanosecond timestamp, labels, and the content of the log line is used to create a unique id. On the other hand, Elasticsearch returns an `_id` field that is unique for the specified index. In such cases, to ensure uniqueness, both the `index name` and `_id` are used to create a unique id.

Refer to [Logs data frame format](#logs-data-frame-format) for more information.

### Log result meta information

> To be implemented in logs data frame as field and/or meta

[Log result meta information]({{< relref "../../../../explore/logs-integration/#log-result-meta-information" >}}) can be used to communicate information about logs result to the user. Information that can be shared with user:

- **Count of received logs vs limit** - It displays the count of received logs compared to the specified limit. Data frames should have a "limit" a meta attribute with the number of requested log lines.

> - **Error**: Displays possible error in your log results. Data frame needs to have "error" meta information.

Author: ```suggestion
- **Error**: Displays possible errors in your log results. Data frames should to have an "error" meta attribute.
```

---
aliases:
  - ../../../plugins/add-support-for-log-queries/
description: Add features to log queries in Explore.
keywords:
  - grafana
  - plugins
  - plugin
  - queries
  - explore queries
  - explore
  - logs
  - log
labels:
  products:
    - enterprise
    - oss
title: Add features to log queries in Explore
weight: 400
---

# Add features to log queries in Explore

> [Explore]({{< relref "../../../../explore" >}}) provides an excellent place for investigating incidents and troubleshooting using log. If the data source produces log results, we highly recommend implementing the following APIs to empower your users to fully utilize the logs UI and its features within Explore.

Reviewer: ```suggestion
[Explore]({{< relref "../../../../explore" >}}) provides an excellent place for investigating incidents and troubleshooting using logs. If the data source produces log results, we highly recommend implementing the following APIs to empower your users to fully utilize the logs UI and its features within Explore.
```

This guide will walk you through the process of adding support for Explore features in a data source plugin through a seamless integration, maximizing the potential for logs analysis. By implementing these APIs, you can enhance the user experience and provide valuable insights through Explore's powerful log investigation capabilities.

## Supported APIs for external plugin developers

### Logs data frame format

> Here is the proposal for how should data frames for log results look like: https://grafana.github.io/dataplane/logs/.
> 

Author: ```suggestion
```

> The data frame should include following fields:

Reviewer 2: Is it exactly the same as the description in data plane page? Maybe it's better not to repeat it, otherwise might be hard to keep it in sync. What do you think?

Author: Good point! Although data plane documentation is currently not published and only part of https://github.com/grafana/dataplane. So when it is published, we can remove this section. 

Also now thinking about it, I will delete the link as it is not an official link.

| Field name     | Field type                              | Info                                                                                                                                                                                                                                       |
| -------------- | --------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |

> | **timestamp**  | `time`                                  | Field with timestamp, non nullable.                                                                                                                                                                                                        |

Reviewer: ```suggestion
| **timestamp**  | `time`                                  | Field with the timestamp, non nullable.                                                                                                                                                                                                        |
```

| **body**       | `string`                                | Field with the content of the log line content, non nullable.                                                                                                                                                                                  |

> | **severity**   | `string`                                | Represents the severity/level of the log line. If no severity field is found, consumers/client will decide the log level. More info about log level can be find [here](https://grafana.com/docs/grafana/latest/explore/logs-integration/). |

Reviewer 3: ```suggestion
| **severity**   | `string`                                | Represents the severity/level of the log line. If no severity field is found, consumers/client will decide the log level. More info about log level can be found [here](https://grafana.com/docs/grafana/latest/explore/logs-integration/). |
```

| **id**         | `string`                                | Unique identifier of the log line.                                                                                                                                                                                                         |

> | **attributes** | `json raw message` (go) or `other` (ts) | This field represent additional attributes of the log line. Other systems may refer to this with other names, for example Loki calls these "labels". Value should be represented with Record<string,any> type in javascript.               |

Reviewer: ```suggestion
| **attributes** | `json raw message` (go) or `other` (ts) | This field represents additional attributes of the log line. Other systems may refer to this with different names, such as "Labels" in Loki. Its value should be represented with Record<string,any> type in javascript.               |
```

Logs data frame's `type` needs to be set to `type: DataFrameType.LogLines` in data frame's meta.

> Example of constructing a logs data frame in `go`:
> 
> > Note: Should be used if you are creating data frames on backend in go code.

Reviewer: ```suggestion
Example of constructing a logs data frame in a `go` backend:
```
I would consider something like this. Or "go backend part of a plugin".

Author: I don't think I understand the comment. Could you please add more context. 

And just to clarify - developers can create data frames in backend and in frontend. Depending on where they are doing it, then go/typescript snippet is relevant.

Author: Oh yeah, makes sense! I'll update.

```go
frame := data.NewFrame(
    "logs",
    data.NewField("timestamp", nil, []time.Time{time.UnixMilli(1645030244810), time.UnixMilli(1645030247027), time.UnixMilli(1645030247027)}),
    data.NewField("body", nil, []string{"message one", "message two", "message three"}),
    data.NewField("severity", nil, []string{"critical", "error", "warning"}),
    data.NewField("id", nil, []string{"xxx-001", "xyz-002", "111-003"}),
    data.NewField("attributes", nil, []json.RawMessage{[]byte(`{}`), []byte(`{"hello":"world"}`), []byte(`{"hello":"world", "foo": 123.45, "bar" :["yellow","red"], "baz" : { "name": "alice" }}`)}),
)

	frame.SetMeta(&data.FrameMeta{
		Type:   data.FrameTypeLogLines,
	})
```

> Example of constructing a logs data frame in `typescript`:

Reviewer: ```suggestion
Example of constructing a logs data frame in `Typescript`:
```

> > Note: Should be used if you are creating data frames on frontend in typescript code.

Reviewer: I would remove or rephrase these notes, as they repeat the same we said before "Example of constructing a logs data frame in Typescript".

Maybe:

```suggestion
> Note: Should only be used if you need to create data frames in the frontend/backend part of the plugin.
```

Author: Same as above

```ts
const result = new MutableDataFrame({
  fields: [
    { name: 'timestamp', type: FieldType.time, values: [1645030244810, 1645030247027, 1645030247027] },
    { name: 'body', type: FieldType.string, values: ['message one', 'message two', 'message three'] },
    { name: 'severity', type: FieldType.string, values: ['critical', 'error', 'warning'] },
    { name: 'id', type: FieldType.string, values: ['xxx-001', 'xyz-002', '111-003'] },
    {
      name: 'attributes',
      type: FieldType.other,
      values: [{}, { hello: 'world' }, { hello: 'world', foo: 123.45, bar: ['yellow', 'red'], baz: { name: 'alice' } }],
    },
  ],
  meta: {
    type: DataFrameType.LogLines,
  },
});
```

### Show log results in Explore's Logs view

> > To be implemented in logs data frame as meta

Reviewer: ```suggestion
> Must be implemented in the data frame as a meta attribute
```

> To ensure that your log results are displayed in interactive Logs view, add data `meta` attribute to `preferredVisualisationType` in your log result data frame.

Reviewer: ```suggestion
To ensure that your log results are displayed in an interactive Logs view, add a `meta` attribute to `preferredVisualisationType` in your log result data frame.
```

> Example of constructing a data frame with specific meta information in `go`:
> 
> > Note: Should be used if you are creating data frames on backend in go code.

Reviewer: ```suggestion
Example of constructing a data frame with specific meta information in `Go`:

> Note: Should be used if you are creating the data frames in the backend.
```

```go
frame.Meta = &data.FrameMeta{

> 	PreferredVisualization: visType,

Reviewer 2: ```suggestion
	PreferredVisualization: "logs",
```

}
```

Example of constructing a data frame with specific meta information in `typescript`:

> Note: Should be used if you are creating data frames on frontend in typescript code.

```ts
const result = new MutableDataFrame({
    fields: [...],
    meta: {
        preferredVisualisationType: 'logs',
    },
});
```

### Color coded log levels

> > To be implemented in logs data frame as field

Reviewer: ```suggestion
> Must be implemented in the data frame as a field
```

> Color coded [log level]({{< relref "../../../../explore/logs-integration/#log-level" >}}) is displayed at the beginning of the log line. It improves log visualization, allowing users to quickly assess the severity of log entries and facilitating effective log analysis and troubleshooting. Log level is determined from `severity` data frame field. In a case when it does not include `severity` field, Grafana tries to evaluate level based on the content of the log line. In a case it is not possible to do so, the log level is set to `unknown`.

Reviewer: ```suggestion
Color coded [log levels]({{< relref "../../../../explore/logs-integration/#log-level" >}}) are displayed at the beginning of each log line. It improves the log visualization by allowing users to quickly assess the severity of log entrie, facilitating log analysis and troubleshooting. The log level is determined from the `severity` field of the data frame. In the case where the `severity` field is not present, Grafana tries to evaluate the level based on the content of the log line. If inferring the log level from the content is not possible, the log level is then set to `unknown`.
```

Refer to [Logs data frame format](#logs-data-frame-format) for more information.

### Copy link to log line

> > To be implemented in logs data frame as field

Reviewer: ```suggestion
> Must be implemented in the data frame as a field
```

> [Copy link to log line]({{< relref "../../../../explore/logs-integration/#copy-link-to-log-line" >}}) is a feature that allows you to generate a link to a specific log line for easy sharing or reference. This feature is supported for data sources that produce log data frames with `id` field.

Reviewer: ```suggestion
[Copy link to log line]({{< relref "../../../../explore/logs-integration/#copy-link-to-log-line" >}}) is a feature that allows you to generate a link to a specific log line for easy sharing and referencing. This feature is supported in data sources that produce log data frames with `id` fields.
```

> In a case where original database does not return `id` field, you can implement one within data source. For instance, in Loki data source, a combination of nanosecond timestamp, labels, and the content of the log line is used to create a unique id. On the other hand, Elasticsearch returns an `_id` field that is unique for the specified index. In such cases, to ensure uniqueness, both the `index name` and `_id` are used to create a unique id.

Reviewer: ```suggestion
In the case where the underlying database does not return an `id` field, you can implement one within the data source. For example, in the Loki data source, a combination of nanosecond timestamp, labels, and the content of the log line is used to create a unique id. On the other hand, Elasticsearch returns an `_id` field that is unique for the specified index. In such cases, to ensure uniqueness, both the `index name` and `_id` are used to create a unique id.
```

Refer to [Logs data frame format](#logs-data-frame-format) for more information.

### Log result meta information

> > To be implemented in logs data frame as field and/or meta

Reviewer: ```suggestion
> Must be implemented in the data frame as a field and/or as meta attribute
```

> [Log result meta information]({{< relref "../../../../explore/logs-integration/#log-result-meta-information" >}}) can be used to communicate information about logs result to the user. Information that can be shared with user:

Reviewer: ```suggestion
[Log result meta information]({{< relref "../../../../explore/logs-integration/#log-result-meta-information" >}}) can be used to communicate information about logs results to the user. The following information can be shared with the user:
```

> - **Count of received logs vs limit** - It displays the count of received logs compared to the specified limit. Data frame needs to have "limit" meta information with number of requested log lines.

Reviewer: ```suggestion
- **Count of received logs vs limit** - It displays the count of received logs compared to the specified limit. Data frames should have a "limit" a meta attribute with the number of requested log lines.
```

> - **Error**: Displays possible error in your log results. Data frame needs to have "error" meta information.

Reviewer: ```suggestion
- **Error**: Displays possible errors in your log results. Data frames need should have an "error" meta attribute.
```

> - **Common labels**: Labels, or attributes that are the same for all displayed log lines are shown as meta information. This feature is supported for data sources that produce log data frames with `attributes` field. Refer to [Logs data frame format](#logs-data-frame-format) for more information.

Reviewer: ```suggestion
- **Common labels**: Labels, or attributes that are the same for all displayed log lines are shown as meta information. This feature is supported for data sources that produce log data frames with an `attributes` field. Refer to [Logs data frame format](#logs-data-frame-format) for more information.
```

> - **Common labels**: Labels, or attributes that are the same for all displayed log lines are shown as meta information. This feature is supported for data sources that produce log data frames with `attributes` field. Refer to [Logs data frame format](#logs-data-frame-format) for more information.

Reviewer 4: i would recommend adding some explanation that for this (common labels) functionality nothing special has to be done by the datasource plugin..
maybe some phrasing like "calculated automatically by the logs panel" or something like that.

Example of constructing a data frame with specific meta information in `go`:

> Note: Should be used if you are creating data frames on backend in go code.

```go
frame.Meta = &data.FrameMeta{
	Custom: map[string]interface{}{
    "limit": 1000,
    "error": "Error information",
  }
}
```

Example of constructing a data frame with specific meta information in `typescript`:

> Note: Should be used if you are creating data frames on frontend in typescript code.

```ts
const result = new MutableDataFrame({
    fields: [...],
    meta: {
        custom: {
          limit: 1000,
          error: "Error information"
        }
    },
});
```

### Highlight searched words

> To be implemented in logs data frame as meta

> Logs visualisation can [highlight specific words or strings]({{< relref "../../../../explore/logs-integration/#highlight-searched-words" >}}) in log entries. This feature is typically utilized to highlight search terms, making it easier for users to locate and focus on relevant information in the logs. For the highlighting to work, search words need to be included in the data frame's meta information

Reviewer: ```suggestion
The kogs visualisation can [highlight specific words or strings]({{< relref "../../../../explore/logs-integration/#highlight-searched-words" >}}) in log entries. This feature is typically utilized for highlighting search terms, making it easier for users to locate and focus on relevant information in the logs. For the highlighting to work, search words need to be included in the data frame's meta information
```

Example of constructing a data frame that includes searchWords in `go`:

> Note: Should be used if you are creating data frames on backend in go code.

```go
frame.Meta = &data.FrameMeta{
	Custom: map[string]interface{}{
    "searchWords": []string{"foo", "bar", "baz"} ,
  }
}
```

Example of constructing a data frame that includes searchWords in `typescript`:

> Note: Should be used if you are creating data frames on frontend in go code.

```ts
const result = new MutableDataFrame({
    fields: [...],
    meta: {
      custom: {
        searchWords: ["foo", "bar", "baz"],
      }
    },
});
```

### Log context

> To be implemented in data source trough implementing DataSourceWithXXXSupport interface

> [Log context]({{< relref "../../../../explore/logs-integration/#log-context" >}}) feature in Explore that enables the display of additional lines of context surrounding a log entry that matches a specific search query. This feature allows users to gain deeper insights into the log data by viewing the log entry in its relevant context. By showing the surrounding log lines, users can better understand the sequence of events and the context in which the log entry occurred, facilitating more comprehensive log analysis and troubleshooting.

Reviewer: ```suggestion
[Log context]({{< relref "../../../../explore/logs-integration/#log-context" >}}) is a feature in Explore that enables the display of additional lines of context surrounding a log entry that matches a specific search query. This feature allows users to gain deeper insights into the log data by viewing the log entry within its relevant context. By showing the surrounding log lines, users can have a better understanding of the sequence of events and the context in which the log entry occurred, facilitating a more comprehensive log analysis and troubleshooting.
```

            status: err.status,
            statusText: err.statusText,
          };
          throw error;
        }),
        // processResultsToDataQueryResponse is a mocked implementation. Be sure to adjust this based on your data source logic.
        switchMap((res) => of(processResultsToDataQueryResponse(res)))
      )
    );
  }

  // Retrieve the context query object for a given log row. This is currently used to open LogContext queries in a split view.
  getLogRowContextQuery(
    row: LogRowModel,
    options?: LogRowContextOptions,
    query?: ExampleQuery
  ): Promise<ExampleQuery | null> {
    // Data source internal implementation that creates context query based on row, options and original query
  }

  // This method can be used to show "context" button based on runtime conditions (for example row model data or plugin settings,...)
  showContextToggle(row?: LogRowModel): boolean {
    // If you want to always show toggle, you can just return true
    if (row && row.searchWords && row.searchWords.length > 0) {
      return true;
    }
  }
}
```

### Filtering of fields using Log details

> To be implemented trough data source method

> Every log line has expandable part called log details that can be opened by clicking on the line. Within the Log details, [Fields]({{< relref "../../../../explore/logs-integration/#fields" >}}) associated with that log entry are shown. If data source implements `modifyQuery?(query: TQuery, action: QueryFixAction): TQuery;` API, then filtering functionality is available for each field. For logs, 2 filtering options are currently available:

Reviewer: ```suggestion
Every log line has an expandable part called "Log details" that can be opened by clicking on the line. Within Log details, [Fields]({{< relref "../../../../explore/logs-integration/#fields" >}}) associated with that log entry are displayed. If the data source implements `modifyQuery?(query: TQuery, action: QueryFixAction): TQuery;` API, then filtering functionality is available for each field. For logs, two filtering options are currently available:
```

- `ADD_FILTER` - to filter for log lines that include selected fields
- `ADD_FILTER_OUT` - to filter for log lines that do not include selected fields

```ts

> export class ExampleDatasource implements DataSourceWithSupplementaryQueriesSupport<ExampleQuery> {

Reviewer 2: I think `DataSourceWithSupplementaryQueriesSupport<ExampleQuery>` is not needed in this example.

Author: Yes, I'll change that, good catch. Copy-paste issue.

        }
        break;
      case 'ADD_FILTER_OUT':
        {
          if (action.options?.key && action.options?.value) {
            // This is a mocked implementation. Be sure to adjust this based on your data source logic.
            queryText = addLabelToQuery(queryText, action.options.key, '!=', action.options.value);
          }
        }
        break;
    }
    return { ...query, query: queryText };
  }
}
```

### Live tailing

> To be implemented trough data source method and enabled in plugin.json

[Live tailing]({{< relref "../../../../explore/logs-integration/#live-tailing" >}}) is a feature that enables real-time log result streaming using Explore. To enable live tailing for your data source, you need to follow these steps:

1. **Enable streaming in plugin.json**: In your data source plugin's plugin.json file, set the `streaming` attribute to `true`. This allows Explore to recognize and enable live tailing controls for your data source.

```json
{
  "type": "datasource",
  "name": "Example",
  "id": "example",
  "logs": true,
  "streaming": true
}
```

2. Ensure that your data source's `query` method can handle queries with `liveStreaming` set to true.

```ts

> export class ExampleDatasource implements DataSourceWithSupplementaryQueriesSupport<ExampleQuery> {

Reviewer 2: Same as above, `DataSourceWithSupplementaryQueriesSupport<ExampleQuery>` is probably not needed here.

Author: Yes, I'll change that, good catch. Copy-paste issue.
